feature:ModulesController

juancmontes opened the pull request:
#### What does it do?
Allow use the MISP API to can query misp_modules.

#### Questions

- [ ] Does it require a DB change?
- [ ] Are you using it in production?
- [X] Does it require a change in the API (PyMISP for example)? it's a new feature for API

#### Release Type:
- [ ] Major
- [X] Minor
- [ ] Patch

juancmontes committed:
new: ModulesQueryAPI

 ModulesQuery controller to can communicate from MISP API to misp_modules

juancmontes committed:
new: ModulesQueryAPI

 ModulesQuery controller to can communicate from MISP API to misp_modules

juancmontes committed:
new: ModulesQueryAPI

 ModulesQuery controller to can communicate from MISP API to misp_modules

juancmontes committed:
new: ModulesQueryAPI

ModulesQuery controller to can communicate from MISP API to misp_modules

juancmontes committed:
new: ModulesQueryAPI

ModulesQuery controller to can communicate from MISP API to misp_modules

juancmontes committed:
new: ModulesQueryAPI

ModulesQuery controller to can communicate from MISP API to misp_modules

juancmontes committed:
new: ModulesQueryAPI

ModulesQuery controller to can communicate from MISP API to misp_modules

juancmontes committed:
new: ModulesQueryAPI

ModulesQuery controller to can communicate from MISP API to misp_modules

juancmontes committed:
Update ModulesQueryController

Support options (credentials) from config.

juancmontes committed:
Update ModulesQueryController

Fix the format of the code

iglocska commented on app/Controller/ModulesQueryController.php:
```diff
@@ -0,0 +1,64 @@
+        if (!empty($options)) $data['config'] = $options;
+
+        // Query
+        $result = $this->Module->queryModuleServer('/query', json_encode($data), true);
```
This allows users that do not have access to a module to query them and is in general not clean. Since we already have all modules via getEnabledModules(), we should check whether the requested module was found in the first place and if not let the user know about it.

juancmontes replied:
Yes, you are right. I will fix it.

iglocska commented on app/Config/routes.php:
```diff
@@ -41,6 +41,10 @@
 	);
+    
+    // ModulesQuery    
+    Router::connect('/modulesQuery/list', array('controller' => 'modulesQuery', 'action' => 'index'));
```
Why not just call the function index? Having a route to index under the list alias seems a bit pointless to me.

juancmontes replied:
This was the first time working with cakephp, so I didnt know if that kind of declarations was mandatory or not. 
I will remove both lines from routes.php and I will rename the list function to index.

iglocska commented on app/Config/routes.php:
```diff
@@ -41,6 +41,10 @@
+    
+    // ModulesQuery    
+    Router::connect('/modulesQuery/list', array('controller' => 'modulesQuery', 'action' => 'index'));
+    Router::connect('/modulesQuery/query', array('controller' => 'modulesQuery', 'action' => 'query'));
```
This one doesn't actually do anything, by default, controller actions are exposed via /[Controller]/[action] - this creates a route for an already existing endpoint.

iglocska commented on app/Controller/Component/ACLComponent.php:
```diff
@@ -191,6 +191,10 @@ class ACLComponent extends Component {
 					'pruneUpdateLogs' => array()
 			),
+            'modulesQuery' => array(
+                'index' => array(),
```
These will restrict the new functionalities to site admins only. Why?

juancmontes replied:
My idea was allow the access only to users with API, and I thought that way to do it.

I will fix it.

iglocska replied:
For users with API access just use this:

`array('perm_auth')`

juancmontes replied:
Ok, thanks

iglocska commented on app/Controller/ModulesQueryController.php:
```diff
@@ -0,0 +1,64 @@
+<?php
+App::uses('AppController', 'Controller');
+class ModulesQueryController extends AppController {
```
This should be called ModulesController.

juancmontes replied:
Ok. That makes sense.

iglocska commented on app/Controller/ModulesQueryController.php:
```diff
@@ -0,0 +1,64 @@
+
+	public function index() {
+        // Initialize models
+		$this->loadModel('Module');
```
Just make sure that you use ModulesController as the controller and that the default table is modules - then the Module model will always be loaded by default.

iglocska commented on app/Controller/ModulesQueryController.php:
```diff
@@ -0,0 +1,64 @@
+        // Query
+        $result = $this->Module->queryModuleServer('/modules', false, true);
+        if ($result) {
+            $result = json_decode('{"error": "Something went wrong, no response from misp_modules."}');
```
Why not `$result = array('error' => 'Something went wrong, no response form misp_modules.'));` ?

juancmontes replied:
I did it to avoid errors on json_encode before send the json. I will try it your way.

iglocska commented on app/Controller/ModulesQueryController.php:
```diff
@@ -0,0 +1,64 @@
+        
+        // Set output
+        $this->set('results', json_encode($result));
+		$this->layout = 'ajax';
```
You can pass JSON responses using the RestResponse component - that way you don't have to create any views

juancmontes replied:
I didnt know it. 

I will use the component and delete the Views.

iglocska replied:
Perfect, thanks!

iglocska reviewed (changes requested):
Thanks a lot, good idea to allow for direct queries to the modules remotely. Could you address the issues in the comments? Let me know if you need any clarification/help with any of them.

Commit:
Merge branch 'ModulesQueryAPI' of https://github.com/juancmontes/MISP into ModulesQueryAPI

Commit:
fix: ModulesController

juancmontes commented:
Fix Views: Removed all Views.
Fix routes.php: not use the file
Fix name: Name changed to ModulesController
Modify query function: Use the method "getEnabledMethod" to validate if the user can query the module.
Modified index function: Only return enabled modules.

juancmontes reviewed (commented):
Fix Views: Removed all Views.
Fix routes.php: not use the file
Fix name: Name changed to ModulesController
Modified query function: Use the method "getEnabledMethod" to validate if the user can query the module.
Modified index function: Only return enabled modules.
Modified returns: Use RestResponse to return json.

Commit:
fix: ModulesController

iglocska commented:
Made some final changes and merged. Changes: 

- query function renamed to query enrichment
- added check for disabled modules and for modules that the current user is not allowed to use
- removed the module config from the index function to avoid exposing API keys / credentials to users
- some formating fixes
